src/OpenTelemetry.Api.ProviderBuilderExtensions/Metrics/MeterProviderServiceCollectionBuilder.cs

```diff
@@ -33,7 +33,8 @@ public MeterProviderServiceCollectionBuilder(IServiceCollection services)
     public MeterProvider? Provider => null;
 
     /// <inheritdoc />
-    public override MeterProviderBuilder AddInstrumentation<TInstrumentation>(Func<TInstrumentation> instrumentationFactory)
+    public override MeterProviderBuilder AddInstrumentation<TInstrumentation>(Func<TInstrumentation?> instrumentationFactory)
```

Comment:
Is it meaningful for instrumentation factories to return null? This code suggests that we expect them to always return a non-null value 

https://github.com/open-telemetry/opentelemetry-dotnet/blob/535c81912b8ed405ae254b43f1498fa248fb2158/src/OpenTelemetry/Metrics/Builder/MeterProviderBuilderSdk.cs#L125-L138

Reply:
Good catch/question!

Short answer: Turns out that assert is based on invalid assumptions.

Long answer:

Today you can do this...

```csharp
using OpenTelemetry;
using OpenTelemetry.Trace;

Console.WriteLine("Starting");

using (var provider = Sdk.CreateTracerProviderBuilder()
    .AddInstrumentation<object>(() =>
    {
        Console.WriteLine("Instrumentation created");
        return null;
    })
    .Build())
{
    Console.WriteLine("Provider created");
}

Console.WriteLine("Ending");
```

Everything will work fine. We don't actually validate that thing returned by the factory is non-null or access it in any way.

What will be logged out is:

```
Starting
Instrumentation created
Provider created
Ending
```

SDK logs look like:

```
2023-09-22T21:55:35.5402710Z:TracerProviderSdk event: '{0}'{Building TracerProvider.}
2023-09-22T21:55:35.5645778Z:TracerProviderSdk event: '{0}'{Sampler added = "OpenTelemetry.Trace.ParentBasedSampler".}
2023-09-22T21:55:35.5650944Z:TracerProviderSdk event: '{0}'{Instrumentations added = "Object".}
2023-09-22T21:55:35.5652432Z:TracerProviderSdk event: '{0}'{TracerProvider built successfully.}
2023-09-22T21:55:35.5660583Z:'{0}' Disposed.{TracerProvider}
```

So `null` is possible today and allowed. There is that `Debug.Assert` in there, but that won't do anything for real code using release builds. That is only validating the repo's code doesn't return a `null` anywhere. Could actually lead us to make the wrong decisions!

It was probably an oversight to allow `null` in the first place, but in order to not break anything I annotated it as supported and made sure the code handles it well. If I go the other way and annotate it as not allowing `null` that could be breaking. We would get warnings in our code where we don't validate things (adding that validation would be breaking), and it may generate warnings for consumers and break their builds. Could also be binary breaking with annotations disapearing, not sure. In the end I decided just to embrace the behavior that we have in place.

Thoughts?

Reply:
PS: Updated tests to make sure this case is captured.